### PR TITLE
[SPARK-45915][SQL] Treat decimal(x, 0) the same as IntegralType in `PromoteStrings`

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

The common type of decimal(x, 0) and string is double. But the common type of int/bigint and string are int/bigint.

This PR updates `PromoteStrings` make the common type of decimal(x, 0) and string is decimal(x, 0).

### Why are the changes needed?

1. Make decimal(x, 0) behave the same as int/bigint in `PromoteStrings`.
2. Reduce one cast in binary comparison so we may use bucket read. For example: `cast(stringCol as double) = cast(decimalCol as double)` vs `cast(stringCol as decimal(x, 0)) = decimalCol`.

### Does this PR introduce _any_ user-facing change?

Yes. The result type of decimal(x, 0) and string is decimal(x, 0) in binary comparison.

### How was this patch tested?

Unit test.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/analyzer-results/typeCoercion/native/binaryComparison.sql.out [hunk 1]
```diff
@@ -1330,7 +1330,7 @@ Project [NOT (cast(cast(null as string) as bigint) = cast(1 as bigint)) AS (NOT
 -- !query
 SELECT cast(1 as decimal(10, 0)) = '1' FROM t
 -- !query analysis
-Project [(cast(cast(1 as decimal(10,0)) as double) = cast(1 as double)) AS (CAST(1 AS DECIMAL(10,0)) = 1)#x]
+Project [(cast(1 as decimal(10,0)) = cast(1 as decimal(10,0))) AS (CAST(1 AS DECIMAL(10,0)) = 1)#x]
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/analyzer-results/typeCoercion/native/binaryComparison.sql.out:1333]: So, is this the main motivation? `cast(1 as double)` -> `cast(1 as decimal(10,0))`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 2]
```diff
@@ -934,8 +934,8 @@ object TypeCoercion extends TypeCoercionBase {
     // There is no proper decimal type we can pick,
     // using double type is the best we can do.
     // See SPARK-22469 for details.
-    case (n: DecimalType, s: StringType) => Some(DoubleType)
-    case (s: StringType, n: DecimalType) => Some(DoubleType)
+    case (DecimalType.Fixed(_, s), _: StringType) if s > 0 => Some(DoubleType)
+    case (_: StringType, DecimalType.Fixed(_, s)) if s > 0 => Some(DoubleType)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:938]: Could you add some description about `s == 0` case after the line 936, please?

**Review by dongjoon-hyun — APPROVED** [review]
Looks reasonable to me.

File: sql/core/src/test/resources/sql-tests/analyzer-results/typeCoercion/native/binaryComparison.sql.out (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/analyzer-results/typeCoercion/native/binaryComparison.sql.out:1333]: Yes.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

**Comment by littlelittlewhite09** [comment]
May I ask a question? Why  the common type of int/bigint and string are int/bigint? For example, `select '37.1' <= 37` will return true. Is that reasonable?

**Comment by wangyum** [comment]
@littlelittlewhite09 Please check the code here: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L937-L938

**Pull request closed without merging**